File: computer3/Fall Term/Solutions/Class09_Nov_06/lvl3_friends.cpp
#include <bits/stdc++.h>
using namespace std;
const int MM = 1e5+2;
int n, nxt[MM]; bool vis[MM];
int main(){
    cin >> n;
    for(int i=1, a, b; i<=n; i++){
        cin >> a >> b; nxt[a] = b;
    }
    while(true){
        int x, y;
        cin >> x >> y;
        if(!x && !y) return 0;
        memset(vis, 0, sizeof(vis));
        int dis = 0, cur = x;
        while(cur != y && !vis[cur]) {
            vis[cur] = true; dis ++;
            cur = nxt[cur];
        }
        if(cur == y) cout << "Yes " << dis-1 << "\n";
        else cout << "No" << "\n";
    }
}

File: computer3/Fall Term/Solutions/Class09_Nov_06/lvl3_friends.java
import java.util.*;
public class Test{
	public static void main(String [] args){
		Scanner in = new Scanner(System.in);
		int [] friend = new int [10000];
		int n = in.nextInt();
		for(int i=0; i<n; i++)
			friend[in.nextInt()] = in.nextInt();
		while(true){
			int x = in.nextInt(), y = in.nextInt();
			if( x==0 && y==0 ) break;
			boolean [] vis = new boolean[10000];
			int dis = 0;
			while(!vis[x] && x != y){
				vis[x] = true;  x = friend[x]; dis++;
			}
			if(x==y) System.out.println("Yes "+(dis-1));
			else System.out.println("No");
		}
		
	}
}